ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java

```diff
@@ -1498,7 +1498,11 @@ protected synchronized int End (boolean reportHeuristics)
             }
             else
             {
-                if (prepare(reportHeuristics) == TwoPhaseOutcome.PREPARE_NOTOK) {
+                int prepareStatus = prepare(reportHeuristics);
+
+                if (prepareStatus == TwoPhaseOutcome.PREPARE_NOTOK
+                        || (prepareStatus == TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED
```

Comment:
One option is to change onePhasePrepare API to return the right status https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java#L2322. That would be a public API change so not advisible for this release but might be something to consider for future.

How about returning TwoPhaseOutcome.PREPARE_NOTOK if the 1PC fails (determined by actionStatus in the place you are currently proposing)? We lose resolution as to the cause of the failure but it minimizes the changeset and the new return code would not actually be used anyway except in your changeset.